Rewrite What is Gradle in a consistent docs voice (#22277)
gradle/gradle: subprojects/docs/src/docs/userguide/what_is_gradle.adoc

// See the License for the specific language governing permissions and
// limitations under the License.

[[gradle_overview]]
[[what_is_gradle]]
= What is Gradle?

Gradle is an open-source https://en.wikipedia.org/wiki/Build_automation[build automation] tool

> flexible enough to build almost any type of software. Gradle bases its design on the following fundamentals:

Author: ```suggestion
flexible enough to build almost any type of software.
Gradle bases its design on the following fundamentals:
```

High performance::
Gradle avoids unnecessary work by only running tasks that need to run because inputs or outputs changed.
Gradle uses various caches to reuse outputs from previous runs. With a shared build cache, you can even reuse outputs across multiple machines.
JVM foundation::
Gradle runs on the JVM. This is a bonus for users familiar with Java, since build logic can use the standard Java APIs.
It also makes it easy to run Gradle on different platforms.
Conventions::
Gradle makes common types of projects easy to build through conventions. Plugins set sensible defaults to keep build scripts minimal.
But these conventions don't limit you: you can configure settings, add your own tasks, and make many other customizations in your builds.
Extensibility::
You can readily extend Gradle to provide your own build logic with custom tasks and plugins.
See Android builds for an example: they add many new build concepts such as flavors and build types.
IDE support::

> Several major IDEs enable imports and interaction with Gradle builds, including Android Studio, IntelliJ IDEA, Eclipse, and NetBeans.

Reviewer: I don't think "enable imports" sounds right, I think the original phrase was better here.

Reviewer: Should we mention VSCode too?  What is the current state of NetBeans support?

Author: ```suggestion
Several major IDEs provide interaction with Gradle builds, including Android Studio, IntelliJ IDEA, Eclipse, VSCode, and NetBeans.
```
Does this sound better to you? As far as I know, there are still supported NetBeans versions that work with supported Gradle versions... but I haven't tested it personally. Open to removing it if you think that it's better that way.

> Gradle can also provides generate the solution files required to load a project into Visual Studio.

Author: ```suggestion
Gradle can also generate the solution files required to load a project into Visual Studio.
```

Gradle can also provides generate the solution files required to load a project into Visual Studio.
Insight::
https://scans.gradle.com/[Build scans] provide extensive information about a build run that you can use to identify build issues. They are particularly good at helping you to identify problems with a build's performance. You can also share build scans with others, which is particularly useful if you need to ask for advice in fixing an issue with the build.

[[five_things]]
== Five things you need to know about Gradle

Gradle is a flexible and powerful build tool that can easily feel intimidating when you first start. However, understanding the following core principles will make Gradle much more approachable and you will become adept with the tool before you know it.
https://scans.gradle.com/[Build scans] provide extensive information about a build that you can use to identify issues.
You can use build scans to identify problems with a build's performance, and even share them for debugging help.

=== 1. Gradle is a general-purpose build tool
== Concepts

> Gradle allows you to build any software, because it makes few assumptions about what you're trying to build or how it should be done. The most notable restriction is that dependency management currently only supports Maven- and Ivy-compatible repositories and the filesystem.

Reviewer: I feel like "Gradle allows you to build any software, because it makes few assumptions about what you're trying to build or how it should be done." was an important point that should remain somewhere.

Author: Agreed. I'll incorporate this elsewhere.

> Projects contain a build script, named `build.gradle` or `build.gradle.kts`.

Reviewer: ```suggestion
Projects contain a build script, which is a file located in the project's root directory named `build.gradle` or `build.gradle.kts`.
```

Reviewer 2: I think I would not use "project's root directory" but simply "project's directory" as the root mention may cause confusion for multi project builds.

> Gradle provides task implementations for many common build sytem goals, like running tests and creating archives.

Reviewer: Maybe "tasks implementing" vs. task implementations, which makes me think this is speaking about OO types.

Author: ```suggestion
Gradle provides tasks that implement many common build system goals, like running tests and creating archives.
```
How does this sound?

> Plugins provide even more task implementations.

Reviewer: ```suggestion
Plugins provide even more types of tasks.
```

=== Plugins

Plugins provide a means of reusing logic and configuration across multiple projects.
With plugins, you can write a task once and use it in multiple builds. Or you can store common configuration,
like logging, dependencies, and version management, in one place. This reduces duplication in
build scripts.

=== Build Phases

Gradle evaluates and executes build scripts in three *build phases* of the <<build_lifecycle#build_lifecycle,Build Lifecycle>>:

Initialization::
Sets up the environment for the build and determine which projects will take part in it.

Configuration::
Constructs and configures the task graph for the build. Determines which tasks need to run and in which order, based on the task the user wants to run.

Execution::
Runs the tasks selected at the end of the configuration phase.

=== Builds

A build is an execution of a collection of tasks in a Gradle project. You instantiate a build with a Gradle command

> via the CLI or an IDE. In the command, you specify a task or tasks. Gradle determines which tasks to run based

Reviewer: ```suggestion
via the CLI or an IDE. In the command, you specify a task or tasks. Gradle determines the complete set of tasks to run based
```

on inter-task dependencies, and runs the smallest set of tasks required to complete the specified task and its dependencies.

== Optimizations

Over the years, Gradle has engineered a number of ways to make your builds faster.
Gradle enables some of these, like the Daemon, by default. You must explicitly enable some other optimizations
through configuration or build logic. This section highlights some of the most important
optimizations to know about when you start using Gradle.

=== Configuration Cache

The configuration phase is a large up-front cost for all Gradle builds. Every time you build your project, Gradle
configures the project based on available inputs like build scripts, settings, and system properties. The
configuration cache allows Gradle to cache the result of this configuration. This can save significant time
in every build. However, the configuration cache is only compatible with certain syntax.

=== Build Cache

Every time you run a build, Gradle runs a set of tasks.
The build cache stores the inputs provided to and outputs generated by those tasks.
When you run Gradle with the build cache enabled, Gradle checks the inputs for each

> task. If the build cache contains output for a repeated set of task inputs, Gradle reuses that output.

Reviewer: ```suggestion
task. If the build cache already contains output for a previously completed task's inputs, Gradle reuses that output immediately, without re-running the task.
```

Gradle can speed up builds significantly by pulling data from the cache instead of running the same task repeatedly.
However, the build cache requires annotated inputs and outputs on tasks.

=== Script Design

Well-designed build scripts consist mostly of
<<authoring_maintainable_build_scripts#sec:avoid_imperative_logic_in_scripts,declarative configuration, rather than imperative logic>>.

> That configuration runs during the configuration phase. Many builds also have task actions

Reviewer: I think `doLast` and `doFirst` are something we might want to de-emphasize in favor or writing custom tasks as fully testable classes.

// See the License for the specific language governing permissions and
// limitations under the License.

[[gradle_overview]]
[[what_is_gradle]]
= What is Gradle?

> Gradle is an open-source https://en.wikipedia.org/wiki/Build_automation[build automation] tool
> flexible enough to build almost any type of software.

Author: ```suggestion
Gradle is an open-source https://en.wikipedia.org/wiki/Build_automation[build automation] tool
flexible enough to build almost any type of software.
This flexibility is the result of Gradle making few assumptions about what you're trying to build or how to build it.
```
@tresat What do you think of this alternate location -- right in the introduction? https://github.com/gradle/gradle/pull/22277#discussion_r986795554

Gradle bases its design on the following fundamentals:

High performance::

> Gradle avoids unnecessary work by only running tasks that need to run because inputs or outputs changed.
> Gradle uses various caches to reuse outputs from previous runs.
> With a shared build cache, you can even reuse outputs across multiple machines.

Reviewer 3: "have changed" reads better to me here, but I can't say if that's more correct.

```suggestion
Gradle avoids unnecessary work by only running tasks that need to do work because inputs or outputs have changed.
Gradle uses various caches to reuse outputs from previous builds.
With a shared build cache, you can even reuse outputs from other machines.
```

JVM foundation::
Gradle runs on the JVM. This is a bonus for users familiar with Java, since build logic can use the standard Java APIs.
It also makes it easy to run Gradle on different platforms.
Conventions::
Gradle makes common types of projects easy to build through conventions. Plugins set sensible defaults to keep build scripts minimal.
But these conventions don't limit you: you can configure settings, add your own tasks, and make many other customizations in your builds.
Extensibility::
You can readily extend Gradle to provide your own build logic with custom tasks and plugins.
See Android builds for an example: they add many new build concepts such as flavors and build types.
IDE support::
Several major IDEs provide interaction with Gradle builds, including Android Studio, IntelliJ IDEA, Eclipse, VSCode, and NetBeans.
Gradle can also generate the solution files required to load a project into Visual Studio.
Insight::

> https://scans.gradle.com/[Build scans] provide extensive information about a build that you can use to identify issues.

Reviewer 3: ```suggestion
https://scans.gradle.com/[Build Scan™] provide extensive information about a build that you can use to identify issues.
```

Author: ```suggestion
https://scans.gradle.com/[Build Scan™] provides extensive information about a build that you can use to identify issues.
```
Good idea. Implementing my own suggestion for subject/verb agreement.

> You can use build scans to identify problems with a build's performance, and even share them for debugging help.

Reviewer 3: ```suggestion
You can use Build Scans to identify problems with a build's performance and even share them for debugging help.
```

> == Concepts

Reviewer 3: Should these sections be linking out to other chapters?

Author: I'd like to do that eventually, but right now we don't have great pages to link to about these concepts -- they're fractured over many pages. As I coalesce those fractured ideas into sections for each concept, I'll add the links here.

It's helpful to know the following concepts before you dive into the details of Gradle.

> === Projects

Reviewer 3: Do we need to define "Build" as a container of projects and other (included) builds?

Author: That's a good point. Do you feel like that isn't expressed well here? I don't *think* this description precludes that idea, but happy to add an extra sentence if you think we should clarify more.

Projects are the things that Gradle builds.

> Projects contain a build script, which is a file located in the project's root directory named `build.gradle` or `build.gradle.kts`.

Reviewer 3: They can be named anything, but these are the default names.

Author: Is that detail necessary to bring up in the "What is Gradle" introduction? If so, I'd suggest this minor change to maintain technical correctness without too much extra verbiage:

```suggestion
Projects contain a build script, which is a file located in the project's root directory usually named `build.gradle` or `build.gradle.kts`.
```

Reviewer 3: 👍

> Builds scripts define tasks, dependencies, plugins, and other configuration for your build.

Reviewer 3: ```suggestion
Builds scripts define tasks, dependencies, plugins, and other configuration for that project.
```

> A single project can contain many subprojects, which can each contain their own subprojects.

Reviewer 3: ```suggestion
A single build can contain one or more projects and each project can contain their own subprojects.
```

> === Tasks

Reviewer 3: I think I would put this section below Plugins

Author: I intentionally placed this section before plugins because IMO tasks are a building block of plugins. But others have raised the point that many introductory users don't write tasks at all -- they just use plugins. So maybe we should put plugins first?

I'm concerned that the concept of plugins is pretty abstract unless you anchor it in something more concrete like tasks. I'll defer to your opinion on this.

Reviewer: I agree with Nate here.  You almost always use a plugin to define tasks, but you can define tasks outside of plugins.  The tasks section seems like it is independent of plugins and should come first.

> Tasks contain the logic that implements builds.
> In most use cases, you'll use or extend existing tasks, like the built-in Java `test` task.
> Gradle provides tasks that implement many common build system goals, like running tests and creating archives.
> Plugins provide even more types of tasks.

Reviewer 3: I'm not sure "implements builds" means anything to a newbie. WDYT?

```suggestion
Tasks contain the logic for executing some work--compiling code, running tests or deploying software.
In most use cases, you'll use existing tasks.
Gradle provides tasks that implement many common build system needs, like the built-in Java `Test` task that can run tests.
Plugins provide even more types of tasks.
```

build scripts.

One last thing: Gradle's <<more_about_tasks#sec:up_to_date_checks,incremental build>> support is robust and reliable, so keep your builds running fast by avoiding the `clean` task unless you actually do want to perform a clean.
=== Build Phases

=== 3. Gradle has several fixed build phases
Gradle evaluates and executes build scripts in three *build phases* of the <<build_lifecycle#build_lifecycle,Build Lifecycle>>:

It's important to understand that Gradle evaluates and executes build scripts in three phases:

 1. Initialization
+
Initialization::
Sets up the environment for the build and determine which projects will take part in it.

 2. Configuration
+
Constructs and configures the task graph for the build and then determines which tasks need to run and in which order, based on the task the user wants to run.
Configuration::
Constructs and configures the task graph for the build. Determines which tasks need to run and in which order, based on the task the user wants to run.

 3. Execution
+
Execution::
Runs the tasks selected at the end of the configuration phase.

These phases form Gradle's <<build_lifecycle#build_lifecycle,Build Lifecycle>>.

[NOTE]

> .Comparison to Apache Maven terminology

Reviewer 3: Are these covered in the migrating from Maven chapter already?

Author: I believe so. Should I add a link to that chapter here, since so many potential users have Maven experience?

Reviewer 3: Yeah, that might be a good idea here

> A build is an execution of a collection of tasks in a Gradle project. You instantiate a build with a Gradle command
> via the CLI or an IDE. In the command, you specify a task or tasks. Gradle determines the complete set of tasks to run based
> on inter-task dependencies, and runs the smallest set of tasks required to complete the specified task and its dependencies.

Reviewer 3: ```suggestion
A build is an execution of a collection of tasks in a Gradle project. You run a build via the CLI or an IDE by specifying task selectors. Gradle configures the build and selects the tasks to run. Gradle runs the smallest complete set of tasks required from the requested tasks and their dependencies.
```

> == Optimizations

Reviewer 3: Is this the right place for this? This chapter is "What is Gradle?".  Should this be in the performance chapter? Should this be under a "Next steps" or linked from Getting Started somehow?

Reviewer 3: And should these sections be linking out to longer form chapters as well?

Author: I mostly meant to bring up important concepts that users will inevitably encounter even in their early days of Gradle usage, so they at least have some basic understanding of ideas like the configuration cache and the build cache. Do you feel like that's a good idea?

Reviewer 3: The Optimizations section just seems out of place for a "What is Gradle?" chapter. 

The way I'm thinking about this is that we have people entering this page from the Getting Started or from The Google.  I think we want this page to just describe what Gradle is (short description, some terminology, some aspirations/vision) and nothing else, except some breadcrumbs to get to something else. 

I think some of the technical vision still belongs here somehow: https://github.com/gradle/gradle/pull/22277#pullrequestreview-1141502330 

I think as a stretch, I can see listing optimizations as part of the reason that Gradle's fast, but I don't think that's as important to mention here. There are lots of other things that are "optimizations" that could be covered here.

I think it could make sense here to call out things that someone might expect from a modern build system:
- Parallel execution
- Incremental execution
- JVM ABI compilation avoidance
- Local and remote caching
- File system watching

This could also provide a natural place to link to Gradle Enterprise and how GE augments some of these things, but that could be added later.

Author: Removed the section -- I don't think it belongs on this page any more.

> in every build. However, the configuration cache is only compatible with certain syntax.

Reviewer 3: Since this isn't production ready yet, we may not want to mention this yet or at least not first.

Author: Considering the configuration cache is already a concern frequently raised in plugin design, blog posts, and elsewhere in the docs, I figured it was a good idea to raise it here. Open to removing it if you don't agree, but I think the cat is out of the bag.

> Every time you run a build, Gradle runs a set of tasks.
> The build cache stores the inputs provided to and outputs generated by those tasks.
> When you run Gradle with the build cache enabled, Gradle checks the inputs for each
> task. If the build cache already contains output for a previously completed task's inputs, Gradle reuses that output immediately, without re-running the task.
> Gradle can speed up builds significantly by pulling data from the cache instead of running the same task repeatedly.
> However, the build cache requires annotated inputs and outputs on tasks.

Reviewer 3: ```suggestion
When you run Gradle with the build cache enabled, Gradle calculates a cache key for the inputs for each
task. If the build cache already contains the outputs for that key, Gradle reuses those outputs immediately without re-running the task.
Enabling the build cache can speed up builds significantly by pulling data from the cache instead of running the same task repeatedly.
```

> That configuration runs during the configuration phase. Many builds also have task actions
> — for example via `doLast {}` and `doFirst {}` blocks — which run during the execution phase.

Author: ```suggestion
That configuration runs during the configuration phase.
```
@tresat does it sound better to remove the `doLast` and `doFirst` references entirely, or should we replace them with something else?

Reviewer: How about: "That configuration runs during the configuration phase.  Many builds also register instances of custom task types providing additional actions that run during the execution phase."

Reviewer: I think references to doLast and doFirst should be completely removed.

> Well-designed build scripts consist mostly of
> <<authoring_maintainable_build_scripts#sec:avoid_imperative_logic_in_scripts,declarative configuration, rather than imperative logic>>.
> That configuration runs during the configuration phase.
> Many builds also register instances of custom task types providing additional actions that run during the execution phase.
> 
> It's easy to view Gradle's build scripts as executable code, because that's what they are.
> But that's an implementation detail: well-designed build scripts describe _what_ steps build the software,
> not _how_ those steps do the work. That's a job for custom task types and plugins.

Reviewer 3: This section seems out of place from the other "Optimizations"

I'm also not sure what we're trying to convey in this section. I like that we're talking about declarative build scripts and how tasks/plugins should hide the implementation details, but I think that message is muddled here. 

```suggestion
Well-designed build scripts consist mostly of
<<authoring_maintainable_build_scripts#sec:avoid_imperative_logic_in_scripts,declarative configuration, rather than imperative logic>>.
That configuration runs during the configuration phase.
Many builds also register instances of custom task types providing additional actions that run during the execution phase.

It's easy to view Gradle's build scripts as executable code, because that's what they are.
But that's an implementation detail: well-designed build scripts describe _what_ steps build the software,
not _how_ those steps do the work. That's a job for custom task types and plugins.
```

Author: Would it make sense to bump this to the top level and rename it "best practices" or something similar? The idea of declarative build scripts is so core to modern idiomatic Gradle that I think it's worth bringing up, but I think you're correct -- it's not really an optimization, it's just the way you should write Gradle logic going forward.

Maven's concept of a build lifecycle is loosely similar to Gradle's <<base_plugin#sec:base_tasks,lifecycle tasks>>.
====

Well-designed build scripts consist mostly of <<authoring_maintainable_build_scripts#sec:avoid_imperative_logic_in_scripts,declarative configuration rather than imperative logic>>. That configuration is understandably evaluated during the configuration phase. Even so, many such builds also have task actions — for example via `doLast {}` and `doFirst {}` blocks — which are evaluated during the execution phase. This is important because code evaluated during the configuration phase won't see changes that happen during the execution phase.

Another important aspect of the configuration phase is that everything involved in it is evaluated _every time the build runs_. That is why it's best practice to <<authoring_maintainable_build_scripts#sec:minimize_logic_executed_configuration_phase,avoid expensive work during the configuration phase>>. https://scans.gradle.com/[Build scans] can help you identify such hotspots, among other things.

=== 4. Gradle is extensible in more ways than one

> It would be great if you could build your project using only the build logic bundled with Gradle, but that's rarely possible. Most builds have some special requirements that mean you need to add custom build logic.

Reviewer 4: I really like these lines. Perhaps we can cut a lot of this content and put it in a section like 
customizing builds?

Author: Just added this information in the "design" section's Extensibility header.

Gradle provides several mechanisms that allow you to extend it, such as:

 * <<custom_tasks#custom_tasks,Custom task types>>.
+
When you want the build to do some work that an existing task can't do, you can simply write your own task type. It's typically best to put the source file for a custom task type in the <<organizing_gradle_projects#sec:build_sources,_buildSrc_>> directory or in a packaged plugin. Then you can use the custom task type just like any of the Gradle-provided ones.

 * Custom task actions.
+

> You can attach custom build logic that executes before or after a task via the link:{groovyDslPath}/org.gradle.api.Task.html#org.gradle.api.Task:doFirst(org.gradle.api.Action)[Task.doFirst()] and link:{groovyDslPath}/org.gradle.api.Task.html#org.gradle.api.Task:doLast(org.gradle.api.Action)[Task.doLast()] methods.

Reviewer 4: glad these are gone

 * <<writing_build_scripts#sec:extra_properties,Extra properties>> on projects and tasks.
+
These allows you to add your own properties to a project or task that you can then use from your own custom actions or any other build logic. Extra properties can even be applied to tasks that aren't explicitly created by you, such as those created by Gradle's core plugins.

 * Custom conventions.
+
Conventions are a powerful way to simplify builds so that users can understand and use them more easily. This can be seen with builds that use standard project structures and naming conventions, such as <<building_java_projects#building_java_projects,Java builds>>. You can write your own plugins that provide conventions — they just need to configure default values for the relevant aspects of a build.

 * <<implementing_gradle_plugins.adoc#modeling_dsl_like_apis,A custom model>>.
+

> Gradle allows you to introduce new concepts into a build beyond tasks, files and dependency configurations. You can see this with most language plugins, which add the concept of <<building_java_projects#sec:java_source_sets,_source sets_>> to a build. Appropriate modeling of a build process can greatly improve a build's ease of use and its efficiency.

Reviewer 4: I like this too. is it going somewhere else? T

Author: Added this into the plugins section!

> You run a build via the CLI or an IDE by specifying task selectors. Gradle configures the build and selects the tasks to run.

Reviewer 4: For a basic page, and since a lot do not use it we should spell out CLI
